File: wing_aeroelastic_analysis.m
% Clear all
clear; clc; close all;

%% Defining initial parameters
c_1 = 0.3;
c_2 = 0.4;
c_r = c_1 + c_2;
c_t = 1.75*c_1;
c_mean = (c_r + c_t)/2;
s = 2;
rho = 1.225;

% Lift distribution function
cl_alpha = @(eta) 2*pi .* sqrt(1 - eta.^2);
% Chord function
c = @(eta) c_r - (c_r - c_t).*eta;
% ec function
ec = @(eta) c(eta) - c_1 - 0.25.*c(eta);

% GJ constants
GJ_1 = 8500;
GJ_2 = 7500;
GJ = @(eta) (GJ_1 * (1 - 0.25.*eta)) .* (eta < 0.5) + ...
            (GJ_2 * (1 - 0.2.*eta)) .* (eta >= 0.5);

%% Part 1: Divergence dynamic pressure convergence study
n_max = 10;

% Cell arrays for mode functions and derivatives
f = cell(n_max, 1);
f_prime = cell(n_max, 1);

% Matrices to store results
q_div = zeros(1, n_max);
error_pct = zeros(1, n_max);

% Initialize convergence flag
converged = false;
n_converged = n_max;  % Initialize to n_max

% Create mode functions and derivatives
for i = 1:n_max
    f{i} = @(eta) i.*eta.^i;              % f_i = i*η^i
    f_prime{i} = @(eta) i.^2 .* eta.^(i-1); % f'_i = i^2*η^(i-1)
end

fprintf('========================================\n');
fprintf('CONVERGENCE STUDY\n');
fprintf('========================================\n');
fprintf('Mode\t\tq_d [N/m²]\t\tError [%%]\n');
fprintf('----\t\t----------\t\t----------\n');

% Try each mode function until convergence
for n = 1:n_max
    % Initialize stiffness matrices (n×n)
    E = zeros(n, n);
    K = zeros(n, n);
    
    % Build E and K matrices
    for i = 1:n
        for j = 1:n

            % E matrix: 
            integrand_E1 = @(eta) (GJ_1 * (1 - 0.25*eta)) .* ...
                                  f_prime{i}(eta) .* f_prime{j}(eta);
            E_part1 = integral(integrand_E1, 0, 0.5);
            
            integrand_E2 = @(eta) (GJ_2 * (1 - 0.2*eta)) .* ...
                                  f_prime{i}(eta) .* f_prime{j}(eta);
            E_part2 = integral(integrand_E2, 0.5, 1.0);
            
            E(i,j) = (s) * (E_part1 + E_part2);
            
            % K matrix: 
            integrand_K = @(eta) c(eta) .* ec(eta) .* cl_alpha(eta) .* ...
                                 f{i}(eta) .* f{j}(eta);
            K(i,j) = (-s) * integral(integrand_K, 0, 1);
        end
    end
    
    % Solve [E] + q[K] = 0
    eigenvalues = eig(E, -K);
    eigenvalues = sort(eigenvalues);
    q_div(n) = eigenvalues(1);
    
    % Display results
    if n == 1
        fprintf('%d\t\t%.2f\t\t--\n', n, q_div(n));
    else
        error_pct(n) = abs((q_div(n) - q_div(n-1)) / q_div(n)) * 100;
        fprintf('%d\t\t%.2f\t\t%.4f\n', n, q_div(n), error_pct(n));
        
        % Check convergence
        if error_pct(n) < 0.1
            fprintf('\n✓ CONVERGED at n=%d modes (error < 0.1%%)\n\n', n);
            converged = true;
            n_converged = n;
            break;
        end
    end
end

% Final results
fprintf('========================================\n');
fprintf('FINAL RESULTS\n');
fprintf('========================================\n');

if converged
    q_d_final = q_div(n_converged);
    fprintf('✓ Converged at: n = %d modes\n', n_converged);
    fprintf('Divergence Dynamic Pressure: q_d = %.2f N/m²\n', q_d_final);
    fprintf('Divergence Speed: U_d = %.2f m/s\n', sqrt(2*abs(q_d_final)/rho));
else
    fprintf('✗ Did NOT converge within %d modes\n', n_max);
    fprintf('Using last value: q_d = %.2f N/m² (n=%d)\n', q_div(n_max), n_max);
    q_d_final = q_div(n_max);
    n_converged = n_max;
end

% Plot convergence
figure('Position', [100, 100, 800, 500]);
plot(1:n_converged, q_div(1:n_converged), 'bo-', 'LineWidth', 2, 'MarkerSize', 8, 'MarkerFaceColor', 'b');
xlabel('Number of Modes', 'FontSize', 12);
ylabel('Divergence Dynamic Pressure q_d [N/m²]', 'FontSize', 12);
title('Convergence of q_d vs Number of Modes', 'FontSize', 14, 'FontWeight', 'bold');
grid on;

if converged
    hold on;
    plot(n_converged, q_d_final, 'ro', 'MarkerSize', 15, 'LineWidth', 3);
    legend('q_d', sprintf('Converged (n=%d)', n_converged), 'Location', 'best');
else
    title('q_d vs Number of Modes (NOT CONVERGED)', 'FontSize', 14, 'FontWeight', 'bold');
end

File: wing_aeroelastic_analysis.py
# -*- coding: utf-8 -*-
"""
Created on Sat Oct 25 23:51:38 2025
AER722 - Project 1
@author: Jawad Rizwan, Peter Stefanov
"""
import math
import sympy as sp
import numpy as np
import scipy.linalg as la
import matplotlib.pyplot as plt

"""
CONSTANTS
--------------------------------------------------
Global variables, can be edited per question part
"""
y = sp.Symbol('y')
eta = sp.Symbol('eta')
k = 0
c1 = 0.3 # m
c2 = 0.4 # m
s = 2 # m - numerical value of span
GJ1 = 8500 # Nm^2/rad
GJ2 = 7500 # Nm^2/rad

"""
BASELINE FORMULAS
--------------------------------------------------
"""

# CLb and CMb are assumed to be constant along span and equal to thin airfoil theory values **GET FROM THIN AIRFOIL THEORY**

cr = c1 + c2 # Root chord length
ct = 1.75*c1 # Tip chord length

#alphaBar = 5-k*eta

"""
FUNCTIONS
--------------------------------------------------
"""

# chord length variation along span
def c(eta, cr, ct): 
    return cr - (cr - ct)*eta

# distance from elastic axis to aerodynamic center
def ec(eta, c1, cr, ct): 
    chord = c(eta, cr, ct)
    return chord - c1 - 0.25*chord

# lift curve slope
def CLa(eta): 
    return 2*math.pi*sp.sqrt(1 - eta**2)
    
# use assumed mode method with as many modes as required in series 1=y, 2=2y^2, 3=3y^3...
def nextmode(n, eta):
    return n*eta**n

def SS (modei, modej, GJ1, GJ2, eta, s): #structural stiffness
    integrand1 = sp.integrate(GJ1*(1 - 0.25*eta)*sp.diff(modei)*sp.diff(modej),(eta, 0, 0.5))
    integrand2 = sp.integrate(GJ2*(1 - 0.2*eta)*sp.diff(modei)*sp.diff(modej),(eta, 0.5, 1))
    return (1/s)*(integrand1 + integrand2)

def AS (modei, modej, eta, s): #aerodynamic stiffness
    # Calculate all parameters inside the function
    chord = c(eta, cr, ct)
    ecfun = ec(eta, c1, cr, ct)
    cla = CLa(eta)
    func = chord*ecfun*cla*modei*modej
    return float(s*(sp.integrate(func,(eta, 0, 1))).evalf())

def error(previous, new):
    return((previous-new)/new*100) # returns percentage error

'''
Question a)
'''

# Max number of modes
nmax = 5

# Array to store dynamic pressures and errors for each mode count
qd = np.zeros(nmax)
modeError = np.zeros(nmax)

# Values at convergence
qdconv = 0
errconv = 0
Kconv = []
Econv = []
Fconv = []

# Finding the first 5 modes (use integer mode numbers to avoid float exponents)
modes = []
for x in range(1, 6):  
    modes.append(nextmode(x, eta))
print(modes)

print("\nMode functions:")
for i, prime in enumerate(modes, 1):
    print(f"Mode {i} function: {prime}")

# Finding the derivatives of the first 5 modes
modePrimes = []
for mode in modes:
    modePrimes.append(sp.diff(mode, eta))

print("\nMode derivatives:")
for i, prime in enumerate(modePrimes, 1):
    print(f"Mode {i} derivative: {prime}")

# Test mode 1
print("\nTest for Mode 1:")
test_ss = float(SS(modes[0], modes[0], GJ1, GJ2, eta, s).evalf())
test_as = float(AS(modes[0], modes[0], eta, s))  # AS already returns float after our previous modification

print("SS value for Mode 1:", test_ss)
print("AS value for Mode 1:", test_as)
mat_ss = np.array([[test_ss]])
mat_as = np.array([[test_as]])

# Calculate just the eigenvalues using numerical values
eigenvalues = (la.eigvals(mat_ss, mat_as)).real  # Note the negative sign before test_as
print("Dynamic Pressure Mode 1:", eigenvalues)

# Calculate the first 5 modes check for convergence
for n in range(5):
    # Create empty stiffness matrices
    E = np.zeros((n+1, n+1))
    K = np.zeros((n+1, n+1))
    
    for i in range(n+1):      
        for j in range(n+1):  
            E[i, j] = float(SS(modes[i], modes[j], GJ1, GJ2, eta, s).evalf())
            K[i, j] = float(AS(modes[i], modes[j], eta, s))
            
    # Calculate divergence dynamic pressure
    eigenvalues = la.eigvals(E, K)
    real_parts = eigenvalues.real
    qd[n] = np.min(real_parts)
    
    # Check if error is less than 0.1% to stop early
    if n > 0:
        modeError[n] = error(qd[n-1], qd[n])
        if abs(modeError[n]) < 0.1:
            print(f"\nConverged at mode count {n+1} with dynamic pressure: {qd[n]}")  
            
            # Save converged values
            qdconv = qd[n]
            errconv = modeError[n]
            Kconv = K
            Econv = E
            
            break

# Print final dynamic pressures for each mode count
print("Final Dynamic Pressures for each mode count:")
for i in range(n+1):
    print(f"Mode {i+1}: {qd[i]} | Error: {modeError[i]}%")
    
# Plot the convergence of dynamic pressure vs number of modes
plt.figure(figsize=(10, 6))

# Create array of mode counts
mode_counts = np.arange(1, n+2)  # n+2 because n is 0-indexed and we want 1,2,3,4,5

# Plot dynamic pressure vs modes
plt.plot(mode_counts, qd[:n+1], 'bo-', linewidth=2, markersize=8, label='Divergence Dynamic Pressure')
if qdconv != 0:
    plt.axhline(y=qdconv, color='r', linestyle='--', linewidth=1.5, label=f'Converged value: {qdconv:.4f} Pa')

plt.xlabel('Number of Modes', fontsize=12)
plt.ylabel('Divergence Dynamic Pressure (Pa)', fontsize=12)
plt.title('Divergence Dynamic Pressure vs Number of Modes', fontsize=14, fontweight='bold')
plt.grid(True, alpha=0.3)
plt.legend()
plt.xticks(mode_counts)

plt.tight_layout()
plt.show()